managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -148,6 +151,10 @@
         return position;
     };
     private final LongPairRangeSet<PositionImpl> individualDeletedMessages;
+
+    // Maintain the indexes deleted status of batch messages that not deleted completely
+    // (ledgerId, entryId) -> deleted indexes
+    private final ConcurrentSkipListMap<PositionImpl, BitSet> batchDeletedIndexes;
```

Comment:
Same comment as I made to client side change. We should consider re-using some of the AckSet implementation in bookkeeper client.